### ISSUE TITLE
Call Tf-Idf Spark function from sparklyr

### ISSUE THREAD

**Issue by aissaelouafi** [rstudio/sparklyr#279]
Can we call the Tf-idf features extraction Spark function from `sparklyr` ? if not can we create the tf-idf matrix based on `dplyr` library ? 
http://spark.apache.org/docs/latest/mllib-feature-extraction.html

**Comment by rvm-xx** [comment]
I tried to replicate the example from the documentation (adapt the path to file to your installation), but got an error:

`documents <- spark_read_csv(sc, name="documents", path = 'data/mllib/kmeans_data.txt',
header = FALSE, delimiter = " ")`
`tf <- sc %>% invoke_new("org.apache.spark.mllib.feature.HashingTF") %>% invoke("transform", documents)`

The error is:
`Error in writeType(con, serdeType) : `
` Unsupported type for serialization tbl_spark`
and applies to documents.

I also tried to convert documents to a dataframe with `spark_dataframe(documents)` function but no better result.

**Comment by aissaelouafi** [comment]
I got exactly the same error message

**Comment by kevinushey** [comment]
The problem with these transformers is simply that they expect data in a form that's not friendly to how R data frames are typically constructed.

As input, the `HashingTF` transformer expects a column with each 'cell' being an array of strings -- typically, with that array being generated by one of the tokenization feature transformers.

For example, the following script will successfully invoke the hashing transformer:

```
library(dplyr)
library(sparklyr)
library(janeaustenr)

sc <- spark_connect(master = "local", version = "2.0.1")

austen     <- austen_books()
austen_tbl <- copy_to(sc, austen, overwrite = TRUE)

sdf <- spark_dataframe(austen_tbl)
tokenized <-  sdf %>% 
  na.omit() %>%
  ft_tokenizer(input.col = "text", output.col = "tokens")

transformed <- invoke_new(sc, "org.apache.spark.ml.feature.HashingTF") %>%
  invoke("setInputCol", "tokens") %>%
  invoke("setOutputCol", "hashed") %>%
  invoke("transform", spark_dataframe(tokenized)) %>%
  sdf_register()

hashed <- collect(transformed)
```

Now, you have a data.frame full of list columns, with some of those list column elements being Spark objects (e.g. SparseVectors).

So, it's not obvious what the best way to export these functions is. Do we just do a low-level export of `ft_hashing_tf()`, and document that you need to call it on a tokenized column? Or do we try to be 'helpful' and provide a easy-to-use wrapper?

**Comment by dsblr** [comment]
@kevinushey Thanks for the solution. While I tried to implement the above solution on a corpus having 3000 docs and 5000 number of words I am getting the following error.
error: Caused by: java.lang.OutOfMemoryError: Java heap space

When I set the following parameters; I got a different error:
Error in invoke_method.spark_shell_connection(sc, TRUE, class, method,  :
  No status is returned. Spark R backend might have failed.

Could you please help me with any solution to resolve the issue.

**Comment by kevinushey** [comment]
@dsblr: sorry, I don't know what could be causing this issue. You should consider filing this separately on the issues page with a reproducible example.

**Comment by dsblr** [comment]
@kevinushey no worries, you guys are doing great job. I have raised an issue https://github.com/rstudio/sparklyr/issues/416 , hope will find some solution for this.

**Comment by dsblr** [comment]
@kevinushey I am able to create the matrix for 100 rows.
But I am unable to get the de-serialized TF_IDF which I am planning to use as a input for ml_lda() to do topic modeling.  Please let me know if is it possible to do so ?
austen_x <- austen[1:100,1:2]
austen_tbl <- copy_to(sc, austen_x, overwrite = TRUE)

sdf <- spark_dataframe(austen_tbl)
tokenized <-  sdf %>% 
  na.omit() %>%
  ft_tokenizer(input.col = "text", output.col = "tokens")


transformed <- invoke_new(sc, "org.apache.spark.ml.feature.HashingTF") %>%
  invoke("setInputCol", "tokens") %>%
  invoke("setOutputCol", "hashed") %>%
  invoke("transform", spark_dataframe(tokenized)) %>%
  sdf_register()

hashed <- collect(transformed)
**Output**
text                book      tokens          hashed
                                                                   <chr>               <chr>      <list>          <list>
1                                                  SENSE AND SENSIBILITY Sense & Sensibility  <list [3]> <dbl [262,144]>
2                                                         by Jane Austen Sense & Sensibility  <list [3]> <dbl [262,144]>

**Comment by dsblr** [comment]
@kevinushey could you please help me to understand how the following code works ?
-> invoke_new(sc, "org.apache.spark.ml.feature.HashingTF") 

I would like to know what is the path(physical path or OS path) of the org.apache.spark.ml.feature.HashingTF  ?

**Comment by kevinushey** [comment]
`invoke_new()` here is calling the constructor of the Java (Scala) class [`org.apache.spark.ml.feature.HashingTF`](https://spark.apache.org/docs/latest/api/scala/#org.apache.spark.ml.feature.HashingTF). Please read the [sparklyr extensions documentation](http://spark.rstudio.com/extensions.html) for more information on how `invoke()` works.

**Comment by dsblr** [comment]
@kevinushey thanks for your response. I have create an constructure using java(scala); I would like to know which file path of os should I copy the java(scala) package, so that I can access the constructure . 
Please help me with the os path.

**Comment by dsblr** [comment]
@kevinushey 
 I have created a JAR for a Scala program and would like to access the functions defined in that scala program from Rstudio using sparklyr::invoke_new

While doing the same, I am getting the following error. Please help me to resolve this issue.
steps:::
1.created a scala program
 `package pkg1

object obj1 {
 def main(args: Array[String]) {
 val pi = 3.14
 def area_of_circle(radius: Double): Double = pi * radius *radius;
 println(area_of_circle(5));
 }
 }
 `
 2) Export the same into a JAR and paste the JAR file called F1
 3) Paste the JAR on the path : SPARK_HOME/jars where other apache jars are present
 4) Execute the following command to access the class object

Error


x <- invoke_new(sc, "pkg1.obj1")
 Error: java.lang.ClassNotFoundException: pkg1.obj1
 at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
 at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
 at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
 at java.lang.Class.forName0(Native Method)
 at java.lang.Class.forName(Class.java:264)
 at sparklyr.StreamHandler$.handleMethodCall(stream.scala:77)
 at sparklyr.StreamHandler$.read(stream.scala:55)
 at sparklyr.BackendHandler.channelRead0(handler.scala:49)
 at sparklyr.BackendHandler.channelRead0(handler.scala:14)
 at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
 at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
 at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:244)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
 at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
 at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131)
 at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
 at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
 at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
 at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
 at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
 at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
 at java.lang.Thread.run(Thread.java:745)

**Comment by kevinushey** [comment]
@javierluraschi do we have a canonical way of launching Spark with a set of user-defined JARs?

**Comment by javierluraschi** [comment]
Yes, one can specify additional jars with the `sparklyr.jars.default` config option.

**Comment by kevinykuo** [comment]
Implemented in 0.7.
